engine/gamesys/src/gamesys/scripts/script_resource.cpp

```diff
@@ -460,6 +460,234 @@ static int GraphicsTextureTypeToImageType(int texturetype)
+    dmGameObject::HCollection collection    = dmGameObject::GetCollection(sender_instance);
+
+    luaL_checktype(L, 2, LUA_TTABLE);
+    uint32_t type        = (uint32_t) CheckTableInteger(L, 2, "type");
+    uint32_t width       = (uint32_t) CheckTableInteger(L, 2, "width");
+    uint32_t height      = (uint32_t) CheckTableInteger(L, 2, "height");
+    uint32_t format      = (uint32_t) CheckTableInteger(L, 2, "format");
+    uint32_t max_mipmaps = (uint32_t) CheckTableInteger(L, 2, "max_mipmaps", 0);
+
+    uint8_t max_mipmaps_actual = dmGraphics::GetMipmapCount(dmMath::Max(width, height));
+
+    if (max_mipmaps > max_mipmaps_actual)
+    {
+        dmLogWarning("Max mipmaps %d requested for texture %s, but max mipmaps supported for size (%d, %d) is %d",
+            max_mipmaps, path_str, width, height, max_mipmaps_actual);
+        max_mipmaps = max_mipmaps_actual;
+    }
+
+    // Max mipmap count is inclusive, so need at least 1
+    max_mipmaps                                        = dmMath::Max((uint32_t) 1, max_mipmaps);
+    uint32_t tex_bpp                                   = dmGraphics::GetTextureFormatBitsPerPixel((dmGraphics::TextureFormat) format);
+    dmGraphics::TextureImage::Type tex_type            = (dmGraphics::TextureImage::Type) GraphicsTextureTypeToImageType(type);
+    dmGraphics::TextureImage::TextureFormat tex_format = (dmGraphics::TextureImage::TextureFormat) GraphicsTextureFormatToImageFormat(format);
+    dmGraphics::TextureImage texture_image             = MakeTextureImage(width, height, max_mipmaps, tex_bpp, tex_type, tex_format);
+
+    dmArray<uint8_t> ddf_buffer;
+    dmDDF::Result ddf_result = dmDDF::SaveMessageToArray(&texture_image, dmGraphics::TextureImage::m_DDFDescriptor, ddf_buffer);
+    assert(ddf_result == dmDDF::RESULT_OK);
+
+    void* resource = 0x0;
+    dmResource::Result res = dmResource::CreateResource(g_ResourceModule.m_Factory, path_str, ddf_buffer.Begin(), ddf_buffer.Size(), &resource);
+
+    DestroyTextureImage(texture_image);
+
+    if (res != dmResource::RESULT_OK)
+    {
+        assert(top == lua_gettop(L));
+        return ReportPathError(L, res, canonical_path_hash);
+    }
+
+    dmGameObject::AddDynamicResourceHash(collection, canonical_path_hash);
+
+    dmScript::PushHash(L, canonical_path_hash);
+    assert((top+1) == lua_gettop(L));
+    return 1;
+}
+
+/*# release a resource
+ * Release a resource.
+ *
+ * @warning This is a potentially dangerous operation, releasing resources currently being used can cause unexpected behaviour.
+ *
+ * @name resource.release
+ *
+ * @param path [type:hash|string] The path to the resource.
+ *
+ */
+static int ReleaseResource(lua_State* L)
+{
+    DM_LUA_STACK_CHECK(L, 0);
+    dmhash_t path_hash = dmScript::CheckHashOrString(L, 1);
+
+    dmResource::SResourceDescriptor* rd = dmResource::FindByHash(g_ResourceModule.m_Factory, path_hash);
+    if (!rd) {
+        luaL_error(L, "Could not get resource: %s", dmHashReverseSafe64(path_hash));
```

Comment:
1. The convention is to "return luaL_error()" even though it's doing a longjump.
2. If you use `DM_LUA_STACK_CHECK`, you also need to use the `return DM_LUA_ERROR("my error");`. This is due to how the scope struct in DM_LUA_STACK_CHECK is implemented.

Reply:
Fixed 👍